File: snippets/trigonometry/arcsin.js
// Educational approximation; radians in/out. Direct angles limited to |x| <= 10000.
const PI = 3.141592653589793;
const MAX_ANGLE = 10000;

function arctangent(x) {
  if (!Number.isFinite(x)) throw new RangeError('Input must be finite.');
  const sign = x < 0 ? -1 : 1;
  x = Math.abs(x);
  const reciprocal = x > 1;
  if (reciprocal) x = 1 / x;
  // Two half-angle reductions make the alternating series converge rapidly.
  for (let k = 0; k < 2; k++) x /= 1 + Math.sqrt(1 + x * x);
  let term = x, sum = x;
  for (let k = 1; k <= 24; k++) { term *= -x * x; sum += term / (2 * k + 1); }
  const angle = 4 * sum;
  return sign * (reciprocal ? PI / 2 - angle : angle);
}

function arcsine(x) {
  if (!Number.isFinite(x) || Math.abs(x) > 1) throw new RangeError('Arcsine requires −1 ≤ x ≤ 1.');
  if (Math.abs(x) === 1) return x * PI / 2;
  return arctangent(x / Math.sqrt((1 - x) * (1 + x)));
}

File: snippets/trigonometry/arcsin.py
# Educational approximation; radians in/out. Direct angles limited to |x| <= 10000.
import math  # sqrt, fmod, isfinite only; no trigonometric calls
PI = 3.141592653589793
MAX_ANGLE = 10000

def arctangent(x):
    if not math.isfinite(x):
        raise ValueError("Input must be finite")
    sign = -1 if x < 0 else 1
    x = abs(x)
    reciprocal = x > 1
    if reciprocal: x = 1 / x
    for _ in range(2): x /= 1 + math.sqrt(1 + x * x)
    term = total = x
    for k in range(1, 25):
        term *= -x * x
        total += term / (2 * k + 1)
    angle = 4 * total
    return sign * (PI / 2 - angle if reciprocal else angle)

def arcsine(x):
    if not math.isfinite(x) or abs(x) > 1:
        raise ValueError("Arcsine requires -1 <= x <= 1")
    if abs(x) == 1: return x * PI / 2
    return arctangent(x / math.sqrt((1 - x) * (1 + x)))
